crates/y-sweet-worker/src/s3_store.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use reqwest::{Client, StatusCode};
+use rusty_s3::{Bucket, Credentials, S3Action};
+use std::{cell::OnceCell, time::Duration};
+use time::OffsetDateTime;
+use y_sweet_core::store::Store;
+
+const PRESIGNED_URL_DURATION_SECONDS: u64 = 60 * 60;
+pub struct S3Store {
+    bucket: Bucket,
+    init_bucket: OnceCell<()>,
+    credentials: Credentials,
+    prefix: Option<String>,
+    presigned_url_duration: Duration,
+}
+
+impl S3Store {
+    pub fn new(
+        region: String,
+        bucket_name: String,
+        prefix: Option<String>,
+        aws_access_key_id: String,
+        aws_secret: String,
+    ) -> Self {
+        let credentials = Credentials::new(aws_access_key_id, aws_secret);
+        let endpoint = format!("https://s3.dualstack.{}.amazonaws.com", region)
+            .parse()
+            .expect("endpoint is a valid Url");
+        let path_style = rusty_s3::UrlStyle::VirtualHost;
+        let bucket = Bucket::new(endpoint, path_style, bucket_name, region)
+            .expect("Url has a valid scheme and host");
+
+        let presigned_url_duration = Duration::from_secs(PRESIGNED_URL_DURATION_SECONDS);
+        S3Store {
+            bucket,
+            init_bucket: OnceCell::new(),
+            credentials,
+            prefix,
+            presigned_url_duration,
+        }
+    }
+
+    async fn init_bucket(&self) -> Result<()> {
+        if self.init_bucket.get().is_none() {
+            let action = self.bucket.create_bucket(&self.credentials);
+            let presigned_url =
+                action.sign_with_time(self.presigned_url_duration, &OffsetDateTime::now_utc());
+            let client = Client::new();
+            let _response = client.put(presigned_url).send().await?;
+            self.init_bucket.set(()).map_err(|_| {
+                anyhow::anyhow!("failed to set init_bucket, this should be impossible")
+            })?;
+        }
+
+        self.init_bucket
+            .get()
+            .ok_or(anyhow::anyhow!(
+                "init_bucket should always be set before get()"
+            ))
+            .map(|_| ())
+    }
+
+    async fn get_inited_bucket(&self) -> Result<&Bucket> {
+        self.init_bucket().await?;
+        Ok(&self.bucket)
+    }
+
+    fn prefixed_key(&self, key: &str) -> String {
+        if let Some(path_prefix) = &self.prefix {
+            format!("{}/{}", path_prefix, key)
+        } else {
+            key.to_string()
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl Store for S3Store {
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        let bucket = self.get_inited_bucket().await?;
+        let prefixed_key = self.prefixed_key(key);
+        let object_get = bucket.get_object(Some(&self.credentials), &prefixed_key);
+        let presigned_url =
+            object_get.sign_with_time(self.presigned_url_duration, &OffsetDateTime::now_utc());
```

Comment:
does `OffsetDateTime::now_utc()` work in workers? I thought it would be a runtime panic because there is no clock.

Reply:
https://developers.cloudflare.com/workers/runtime-apis/webassembly/rust/crates/

ye time works, It compiles down to the js date api.